src/lib/profiles/data-management/Current/CommandSender.h

```diff
@@ -0,0 +1,261 @@
+ *
+ * ### ResponseReceived
+ *
+ * Receiving a Response implies that the Command recipient handled the Command successfully. The
+ * Response may contain a payload or it may not. If the applcation desires to keep the packet
+ * buffer, it may call ExchangeContext::AddRef() to increment the ref count.
+ *
+ */
+class CommandSender
+{
+public:
+    enum EventType
+    {
+        kEvent_CommunicationError                 = 1,        // All errors during communication (transmission error, failure to get a response timeout) will be communicated through this event
+        kEvent_InProgressReceived                 = 2,        // Receipt of an 'in progress' message
+        kEvent_StatusReportReceived               = 3,        // On receipt of a status report
+        kEvent_ResponseReceived                   = 4,        // On receipt of a command response
+
+        kEvent_DefaultCheck                         = 100,    //< Used to verify correct default event handling in the application.
+    };
+
+    // Data returned to the sender by the recipient of the custom command.
+    struct InEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+
+        union {
+            struct
+            {
+                Profiles::StatusReporting::StatusReport *statusReport;
+            } StatusReportReceived;
+
+            /*
+             * All communication errors, including transport errors on transmission
+             * as well as failure to receive a response (WEAVE_ERROR_TIMEOUT).
+             */
+            struct
+            {
+                WEAVE_ERROR error;
+            } CommunicationError;
+
+            /* On receipt of a command response, both a reader positioned at the
+             * payload as well as a pointer to the packet buffer is provided. If the applicaton
+             * desires to hold on to the buffer, it can do so by incrementing the ref count on the buffer
+             */
+            struct
+            {
+                uint64_t traitDataVersion;
+                nl::Weave::TLV::TLVReader *reader;
+                nl::Weave::PacketBuffer *packetBuf;
+            } ResponseReceived;
+        };
+    };
+
+    // Data returned back to a CommandSender object from an EventCallback
+    // function.
+    struct OutEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+        bool defaultHandlerCalled;
+    };
+
+    typedef void (*EventCallback)(void * const aAppState, EventType aEvent, const InEventParam &aInParam, OutEventParam &aOutEventParam);
+
+    struct SendParams {
+        WEAVE_ERROR PopulateTraitPath(TraitCatalogBase<TraitDataSink> *aCatalog, TraitDataSink *aSink, uint32_t aCommandType);
+
+        TraitDataSink *Sink;
+
+        ResourceIdentifier ResourceId;
+        uint32_t ProfileId;
+        SchemaVersionRange VersionRange;
+        uint64_t InstanceId;
+        uint32_t CommandType;
+
+        uint8_t Flags;
```

Comment:
@mrjerryjohns it might be good to provide some explanation of these fields. Which ones are required to set?

Reply:
Good idea.

ResourceId, ProfileId, CommandType should be populated. InstanceId, Flags and VersionRange default to 0 if unspecified.

Perhaps having two constructors, one with each argument explicitly specified, and another using the TraitCatalog approach to discover some of those parameters (i.e rename the PopulateTraitPath to be another overload constructor) might be a good idea.

@robszewczyk thoughts?

Reply:
Given that this structure is part of the public API, all fields should be documented, including describing when and how they should be set.

As far as I can see, this structure is only used as a means to avoid having a long list of arguments to SendCommand().  As such, I wouldn't expect it to have a constructor.